synchronizer/l1_rollup_info_producer.go

```
@@ -0,0 +1,401 @@
+		outgoingChannel:                      outgoingChannel,
+		statistics:                           newRollupInfoProducerStatistics(startingBlockNumber),
+		status:                               producerIdle,
+		cfg:                                  cfg,
+	}
+	return &result
+}
+
+// TDOO: There is no min/max function in golang??
+func min[T constraints.Ordered](a, b T) T {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+func max[T constraints.Ordered](a, b T) T {
+	if a > b {
+		return a
+	}
+	return b
+}
+
+func (l *l1RollupInfoProducer) reset(startingBlockNumber uint64) {
+	log.Infof("producer: Reset L1 sync process to blockNumber %d", startingBlockNumber)
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+	log.Debugf("producer: Reset(%d): context cancel", startingBlockNumber)
+	l.cancelCtx()
+	l.ctx, l.cancelCtx = context.WithCancel(l.ctxParent)
+	log.Debugf("producer: Reset(%d): syncStatus.reset", startingBlockNumber)
+	l.syncStatus.reset(startingBlockNumber)
+	l.statistics.reset(startingBlockNumber)
+	log.Debugf("producer: Reset(%d): stop workers (%s)", startingBlockNumber, l.workers.toString())
+	l.workers.stop()
+	// Empty pending rollupinfos
+	log.Debugf("producer: Reset(%d): emptyChannel", startingBlockNumber)
+	l.emptyChannel()
+	log.Debugf("producer: Reset(%d): reset Filter", startingBlockNumber)
+	l.filterToSendOrdererResultsToConsumer.Reset(startingBlockNumber)
+	l.status = producerIdle
+	log.Debugf("producer: Reset(%d): reset done!", startingBlockNumber)
+}
+
+func (l *l1RollupInfoProducer) emptyChannel() {
+	for len(l.outgoingChannel) > 0 {
+		<-l.outgoingChannel
+	}
+}
+
+// verify: test params and status without if not allowModify avoid doing connection or modification of objects
+func (l *l1RollupInfoProducer) verify() error {
+	return l.syncStatus.verify()
+}
+
+func (l *l1RollupInfoProducer) initialize() error {
+	err := l.verify()
+	if err != nil {
+		log.Fatal(err)
+	}
```

If you returns an error in this function why you just log a fatal instead of return the error? We should return the error if possible and try to avoid fatals as much as possible

---

changed!